### DISCUSSION TITLE
can the preconditioner be dynamically changed during a simulation?

### DISCUSSION THREAD

**Discussion by WilkAndy** [idaholab/moose#22692] [Q&A General] [answered]
Hi everyone,

Sometimes my simulations run nicely until a certain time when they slow down due to some physical occurrence, like applying a new load or flux to the system.  In these cases, it may be advantageous to use a different preconditioner, just for those few timesteps around the physical occurrence.  For instance, usually i could use a "weak/cheap" preconditioner like ILU, but around the difficult times, I'd like to use LU.

Is it possible to instruct MOOSE to use different preconditioners at different nonlinear steps (or time steps), depending on some criterion?   Something like:

```
if previous_linear_iterations < 999:
  preconditioner = ilu
else:
  preconditioner = lu
```

**Comment by GiudGiud** [comment, marked as answer]
As far as I know we do not support that.
Even setting the convergence criterion to be time dependent would be super helpful in my mind.

Tagging @friedmud and @lindsayad for awareness and comments.

**Reply by WilkAndy** [reply]
Thanks @GiudGiud .   An alternative could be something like:

```
try:
  ilu
except SolveFailed:
  try:
    hypre
  except SolveFailed:
    try:
      lu
```

**Reply by GiudGiud** [reply]
that could be an alternative strategy to provide more resilience, though you often have to pass options in addition to this, so maybe you d want multiple Preconditioning blocks in the input file. 

I think cutting the time step is a more reasonable choice in the work I do, but I can see that for others.

**Reply by WilkAndy** [reply]
Yea, i agree about cutting the timestep - it often works.  However, i would bet the changing the preconditioner also works in many cases, but we haven't had the flexibility to experiment with that approach.  Cutting the timestep usually helps make the problem more linear, but that doesn't necessarily mean easier for the linear solver.

**Reply by GiudGiud** [reply]
To me it also reduces the distance the nonlinear solver has to travel as well.
for flow in the NS module there s very few PC that would work without expert settings.

**Reply by lindsayad** [reply]
Strictly speaking you could do this in a custom `Executioner` but yea we do not have the ability to do this easily from, say, an input file. However, with @reverendbedford's recent changes to allow keeping around the SNES context between timesteps, I think implementing something like this would not be all that difficult. @WilkAndy feel free to create an issue for this if we don't have one already. I don't know when someone from the MOOSE team would be able to take it on, but maybe an external developer would
